packs/BP/scripts/network.ts

```diff
@@ -89,123 +89,122 @@ export class MachineNetwork extends DestroyableObject {
-          }
-
-          const amount = getMachineStorage(block, queuedSend.type);
-
-          if (amount >= definition.maxStorage) {
-            yield;
-            continue;
-          }
-
-          targetsArr.push({
-            block,
-            amount,
-            definition,
-          });
-
-          yield;
+    // Calculate the amount of each type that is avaliable to send around.
+    const distribution: Record<string, number> = {};
+
+    this.sendQueue.forEach(send => {
+      const newValue = distribution[send.type] ?? 0;
+      distribution[send.type] = newValue + send.amount;
+    });
+
+    // Clear queue for next time.
+    this.sendQueue.splice(0, this.sendQueue.length);
+    const typesToDistribute = Object.keys(distribution);
+
+    interface Consumer { normalPriority: Block[], lowPriority: Block[] };
+
+    // initialize consumers keys. 
+    const consumers = typesToDistribute.reduce<Record<string, Consumer>>((acc, key) => {
+      acc[key] = { lowPriority: [], normalPriority: [] };
+      return acc;
+    }, {});
+
+    // find and filter connections into their consumer groups.
+    this.connections.machines.forEach(machine => {
+      const machineTags = machine.getTags();
+      const isLowPriority = machineTags.includes("fluffyalien_energisticscore:low_priority_consumer");
+      const allowsAny = machineTags.includes("fluffyalien_energisticscore:consumer._any");
+
+      // Check machine tags and sort into appropriate groups.
+      Object.keys(consumers).forEach(consumerType => {
+        const allowsType = allowsAny || machineTags.includes(`fluffyalien_energisticscore:consumer.${consumerType}`);
+        if (!allowsType) return;
+        
+        if (isLowPriority) consumers[consumerType].lowPriority.push(machine);
+        else consumers[consumerType].normalPriority.push(machine);
+      });
+    });
+
+    // send each machine its share of the pool.
+    for (const type of typesToDistribute) {
+      const machines = consumers[type];
+      const numMachines = machines.lowPriority.length + machines.normalPriority.length;
+      if (numMachines === 0) continue;
+
+      let budget = distribution[type];
+
+      // Give each machine in the normal priority an equal split of the budget
+      // Machines can consume less than they're offered in which case the savings are given to further machines.
+      for (let i = 0; i < machines.normalPriority.length; i++) {
+        const machine = machines.normalPriority[i];
+        const budgetAllocation = budget / (machines.normalPriority.length - i);
+        const currentStored = getMachineStorage(machine, type);
+        const machineDef = getMachineRegistration(machine);
+        
+        let amountToAllocate: number = Math.min(budgetAllocation, machineDef.maxStorage - currentStored);
+
+        for (const result of this.sendMachineAllocation(machine, machineDef, type, amountToAllocate)) {
+          amountToAllocate = result ?? amountToAllocate;
```

Comment:
`result` is always `undefined` here as `for..of` does not get the return value of a generator

this also does not `yield` so it is synchronous and it may run for up to 20 ticks before `invokeScriptEvent` throws

my suggestion is to make `sendMachineAllocation` an asynchronous function and do the `while (!completed) yield;` in `send` instead of in `sendMachineAllocation`

Reply:
done